Extend _save_for_mobile and _load_for_mobile to support flatbuffer format; Default format is pickle

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74209

Differential Revision: [D34805092](https://our.internmc.facebook.com/intern/diff/D34805092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34805092/)!

pavithranrao committed:
Change buck targets to support only pickle and pickle + flatbuffer for migration

```lang=mermaid
graph TD;
    torch_core-->torch_mobile_deserialize;

    torch_mobile_core-->torch_mobile_deserialize;

    jit_module_saving-->torch_core;
    jit_module_saving-->torch_mobile_core;

    torch_mobile_deserialize-->caffe2_serialize;
    torch_mobile_deserialize-->torch_mobile_module;

    caffe2_serialize-->miniz;

    flatbuffer_loader-->mobile_bytecode;
    flatbuffer_serializer-->mobile_bytecode;

    mobile_bytecode-->flatbuffer_2.0;

    torch_mobile_deserialize_pickle_and_flatbuffer-->|new| flatbuffer_loader;
    torch_mobile_deserialize_pickle_and_flatbuffer-->|new| torch_mobile_deserialize;
    torch_mobile_core_pickle_and_flatbuffer-->|new| torch_mobile_deserialize_pickle_and_flatbuffer;
    torch_core_pickle_and_flatbuffer-->|new| torch_mobile_deserialize_pickle_and_flatbuffer;

    jit_module_saving_pickle_and_flatbuffer-->|new| torch_core_pickle_and_flatbuffer;
    jit_module_saving_pickle_and_flatbuffer-->|new| torch_mobile_core_pickle_and_flatbuffer;

    flatbuffer_serializer-->torch_mobile_module;

    jit_module_saving_pickle_and_flatbuffer-->|new|jit_module_saving;
    jit_module_saving_pickle_and_flatbuffer-->|new|flatbuffer_serializer;

    flatbuffer_loader-->torch_mobile_module;
```

Differential Revision: [D34805092](https://our.internmc.facebook.com/intern/diff/D34805092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34805092/)!

[ghstack-poisoned]

qihqi commented on tools/build_variables.bzl:
```diff
@@ -622,6 +622,7 @@ libtorch_extra_sources = libtorch_core_jit_sources + [
     "torch/csrc/jit/serialization/export.cpp",
     "torch/csrc/jit/serialization/export_bytecode.cpp",
     "torch/csrc/jit/serialization/export_module.cpp",
+    # "torch/csrc/jit/serialization/flatbuffer_serializer.cpp",
```
remove comment

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.h:
```diff
@@ -52,10 +52,24 @@ TORCH_API mobile::Module load_mobile_module_from_file(
     const std::string& filename,
     c10::optional<at::Device> device = c10::nullopt);
 
+TORCH_API mobile::Module load_mobile_module_from_stream(
```
these functions seems better suited to be in import.h

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -591,7 +592,66 @@ mobile::Module load_mobile_module_from_file(
+  in.seekg(orig_pos, in.beg);
+
+  // read stream
+  std::shared_ptr<char> data(static_cast<char*>(malloc(size)), free); // NOLINT
```
use aligned_alloc on x86 it's fine but I think on 32 bit arm it ll give some errors if not aligned to 16.

qihqi commented on torch/csrc/jit/mobile/flatbuffer_loader.cpp:
```diff
@@ -591,7 +592,66 @@ mobile::Module load_mobile_module_from_file(
+  TORCH_CHECK(
+      mobile::serialization::ModuleBufferHasIdentifier(data.get()),
+      "Format error");
+  auto* flatbuffer_module = mobile::serialization::GetMutableModule(data.get());
```
I'd just inline the 3 lines here to the switch statement in import.cpp.

qihqi commented on torch/csrc/jit/serialization/flatbuffer_serializer.cpp:
```diff
@@ -362,7 +362,7 @@ flatbuffers::DetachedBuffer FlatbufferSerializer::serializeModule(
       storage_data_offset,
       fbb.CreateVector(obj_types_offset_));
-  fbb.Finish(mod);
+  FinishModuleBuffer(fbb, mod);
```
rebase?

qihqi commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -788,44 +791,99 @@ SerializationStorageContext& ScriptModuleSerializer::storage_context() {
+  }
 }
 
 void ExportModule(
```
this function can call the ExportModule that takes function below.

iseeyuan commented on test/cpp/jit/test_flatbuffer.cpp:
```diff
@@ -153,16 +153,30 @@ TEST(FlatbufferTest, ExtraFiles) {
   extra_files["mobile_info.json"] = "{\"key\": 23}";
 
+  std::unordered_map<std::string, std::string> loaded_extra_files;
+#if defined FB_XPLAT_BUILD
```
Why it's only for FB_XPLAT_BUILD?

pavithranrao replied:
fixed this

iseeyuan commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -536,29 +540,89 @@ mobile::Module _load_for_mobile(
+      auto module = _load_for_mobile(std::move(rai), device, extra_files);
+      return module;
+    }
+#if defined(ENABLE_FLATBUFFER)
```
I think we use the magic byte to differentiate. Is the macro used for internal targets that depends on different version of flatbuffers?

pavithranrao replied:
yes; If we build doesn't bring in flatbuffer and tries to load a flatbuffer module we would want to display this error letting the users build with appropriate flags/target deps.

iseeyuan reviewed (commented):
Thanks @pavithranrao 
I saw two macros used in this PR. Could you confirm they are actually needed? If so, could you add comments?

pavithranrao committed:
Update on "Extend _save_for_mobile and _load_for_mobile to support flatbuffer format; Default format is pickle"


Differential Revision: [D34805092](https://our.internmc.facebook.com/intern/diff/D34805092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34805092/)!

[ghstack-poisoned]

pavithranrao committed:
Update on "Extend _save_for_mobile and _load_for_mobile to support flatbuffer format; Default format is pickle"


Differential Revision: [D34805092](https://our.internmc.facebook.com/intern/diff/D34805092/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34805092/)!

[ghstack-poisoned]

qihqi commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -536,29 +540,110 @@ mobile::Module _load_for_mobile(
+      std::tie(data, size) = get_stream_content(in);
+      auto* flatbuffer_module =
+          mobile::serialization::GetMutableModule(data.get());
+      mobile::Module m = initialize_mobile_module(flatbuffer_module);
```
Need to attach data to `m` with`m.set_delete_memory`

qihqi commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -536,29 +540,110 @@ mobile::Module _load_for_mobile(
-  std::unique_ptr<FileAdapter> rai = std::make_unique<FileAdapter>(filename);
-  auto module = _load_for_mobile(std::move(rai), device, extra_files);
-  return module;
+  auto format = getFileFormat(filename);
```
this function and the one in line 661 has has code-dup. Make one call another or refactor common bits

qihqi commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -788,44 +791,99 @@ SerializationStorageContext& ScriptModuleSerializer::storage_context() {
+  };
+  if (use_flatbuffer) {
+#if defined(ENABLE_FLATBUFFER)
+    save_mobile_module_to(
```
this function won't put JIT source into flatbuffer right? is that the intent?

iseeyuan reviewed (approved):
LGTM. Please make sure all CIs pass, and add size estimates as mentioned my internal comment.

malfet commented:
Reverting as it broke Windows builds on master (same as on PR), see https://github.com/pytorch/pytorch/commit/284b2b713592728c362d2d9a2813e021895e89eb
